File: 49_trie_demo.py
"""
Task 49: Trie insert/search demonstration.
Reads N, then N words to insert, then a query word and prints YES/NO if present.
"""
import sys

class TrieNode:
    def __init__(self):
        self.next = {}
        self.end = False

class Trie:
    def __init__(self): self.root = TrieNode()
    def insert(self, word):
        node = self.root
        for ch in word:
            ch = ch.lower()
            if ch not in node.next: node.next[ch] = TrieNode()
            node = node.next[ch]
        node.end = True
    def search(self, word):
        node = self.root
        for ch in word:
            ch = ch.lower()
            if ch not in node.next: return False
            node = node.next[ch]
        return node.end

if __name__ == '__main__':
    data = sys.stdin.read().split()
    if not data: sys.exit(0)
    it = iter(data)
    n = int(next(it))
    trie = Trie()
    for _ in range(n): trie.insert(next(it))
    q = next(it)
    print('YES' if trie.search(q) else 'NO')

# Time Complexity: O(L) per op
# Space Complexity: O(sum lengths)

File: 49_trie_demo.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 49: Trie insert/search demonstration.
Reads N, then N words to insert, then a query word and prints YES/NO if present.
*/

struct TrieNode
{
    array<int, 26> next;
    bool end = false;
    TrieNode() { next.fill(-1); }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int N;
    if (!(cin >> N))
        return 0;
    string w;
    vector<TrieNode> trie(1);
    for (int i = 0; i < N; ++i)
    {
        cin >> w;
        int v = 0;
        for (char c : w)
        {
            int idx = tolower(c) - 'a';
            if (trie[v].next[idx] == -1)
            {
                trie[v].next[idx] = trie.size();
                trie.emplace_back();
            }
            v = trie[v].next[idx];
        }
        trie[v].end = true;
    }
    string q;
    cin >> q;
    int v = 0;
    bool ok = true;
    for (char c : q)
    {
        int idx = tolower(c) - 'a';
        if (trie[v].next[idx] == -1)
        {
            ok = false;
            break;
        }
        v = trie[v].next[idx];
    }
    cout << (ok && trie[v].end ? "YES" : "NO") << "\n";
}

// Time Complexity: insert/search O(length)
// Space Complexity: O(sum of lengths * alphabet)
